### PR TITLE
Post: Feedback on a live-coding to discover Go language

### PR DESCRIPTION
**Pull request by eko**
Progression:

- [X] French version
- [x] English version

### CODE REVIEW

File: _posts/2017-08-23-feedback-on-a-live-coding-to-discover-go-language.md [hunk 1]
```diff
@@ -0,0 +1,367 @@
...
+Conclusion
+----------
+
+Before any conclusion, the open-source code of this library is available here: [https://github.com/unikorp/workmq](https://github.com/unikorp/workmq){:target="_blank"}.
+
+I had two main objectives for this live-coding/workshio session:
+* Allow to my participants to write a complete and functional open-source library in three hours,
+* Allow to my participants to discover most of the Go language features and concepts so that they are able to develop a Go library or application by themselves the day after.
+
+I think the contract is fulfilled with this workshop and I hop it will be useful to you to discover Go language or to let your colleagues know it.
```
Comment (ch3ric) [inline _posts/2017-08-23-feedback-on-a-live-coding-to-discover-go-language.md]: Hope

File: _posts/2017-08-23-feedback-on-a-live-coding-to-discover-go-language.md [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+
+Also, do not hesitate to contact Eleven Labs or contact me directly if you want to organize workshop sessions on web technologies.
```
Comment (ch3ric) [inline _posts/2017-08-23-feedback-on-a-live-coding-to-discover-go-language.md]: :+1:

**Comment by eko** [comment]
Thanks for your review @ch3ric! 🤝
Your feedbacks has been fixed.

File: _posts/2017-08-23-feedback-on-a-live-coding-to-discover-go-language.md [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+		w.Processor(w, message)
+	}
+}
+```
+
+First thing important to explain as soon as you will encounter the case is the `func (w *Worker) Process() {` notation that will allow the `Process()` method to be called on a `Workmq` struct instance type only.
+
+Next, another interesting thing to explain here is the `channel` notation:
+* `<-chan`: indicates that the channel will be used for read only,
+* `chan<-`: indicated that the channel will be used for receiving data only.
```
Comment (ch3ric) [inline _posts/2017-08-23-feedback-on-a-live-coding-to-discover-go-language.md]: indicates

File: _posts/2017-08-23-feedback-on-a-live-coding-to-discover-go-language.md [hunk 4]
```diff
@@ -0,0 +1,367 @@
...
+
+Finally, you can also take a tour to explain `for` loops and its notations coupling with `range` keyword.
+
+Processors (part of core)
+-------------------------
+
+Nothing really new on this part additionnally to workers when declaring our processors so I used this part to present the error handling/return and also keywords and notation for manipulating maps:
```
Comment (ch3ric) [inline _posts/2017-08-23-feedback-on-a-live-coding-to-discover-go-language.md]: for manipulating -> to manipulate?

File: _posts/2017-08-23-retour-sur-un-live-coding-de-decouverte-du-langage-go.md [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+    - worker
+    - workshop
+cover: /assets/2017-08-23-retour-sur-un-live-coding-de-decouverte-du-langage-go/cover.jpg
+---
+
+Cet article fait suite à un workshop / live-coding que j'ai eu l'occasion d'organiser chez Eleven Labs pour une initiation au langage Go.
+
+Les workshops sont pour moi le meilleur moyen d'entrer dans un sujet technique que l'on ne maîtrise pas encore car il permet vraiment la pratique du sujet et en plus d'être accompagné par une personne (pas forcémment experte sur le sujet) qui a déjà pratiquée le sujet, qui a travaillée son workshop et qui donc saura vous donner les billes pour commencer.
+
+Définission du sujet
```
Comment (ch3ric) [inline _posts/2017-08-23-retour-sur-un-live-coding-de-decouverte-du-langage-go.md]: Définition

File: _posts/2017-08-23-retour-sur-un-live-coding-de-decouverte-du-langage-go.md [hunk 6]
```diff
@@ -0,0 +1,366 @@
...
+![WorkMQ Schema](/assets/2017-08-23-retour-sur-un-live-coding-de-decouverte-du-langage-go/schema.jpg)
+
+Comme vous pouvez le voir sur ce diagramme, nous avons ici quatre `Queues` de définies et chacune d'entre elles dispose de trois `Workers`.
+
+Notre librairie (`WorkMQ`, ici le point central), fournira un [Channel (au sens du langage Go)](https://golang.org/ref/spec#Channel_types){:target="_blank"} dans lequel seront stockés les messages de la queue correspondante qui seront dépilés par le premier worker disponible.
+
+Configuration
+-------------
+
+Loin que ce soit le meilleur moyen de gérer la configuration d'une application, le fichier `json` reste une des manières simple et qui permet d'écrire les premières lignes de Go tout en permettant de comprendre les principes de base du langage.
```
Comment (ch3ric) [inline _posts/2017-08-23-retour-sur-un-live-coding-de-decouverte-du-langage-go.md]: Loin que ce soit -> Loin d'être

File: _posts/2017-08-23-retour-sur-un-live-coding-de-decouverte-du-langage-go.md [hunk 7]
```diff
@@ -0,0 +1,366 @@
...
+
+En effet, pour lire la configuration (écrite donc dans un fichier JSON) et la transformer sous forme de `struct` (structures) Go, nous avons ainsi commencés par définir la structure de données de notre configuration :
```
Comment (ch3ric) [inline _posts/2017-08-23-retour-sur-un-live-coding-de-decouverte-du-langage-go.md]: commencé

File: _posts/2017-08-23-retour-sur-un-live-coding-de-decouverte-du-langage-go.md [hunk 8]
```diff
@@ -0,0 +1,366 @@
...
+    }
+  }
+}
+```
+
+Simple et efficace, celle-ci nous permet de définir les ports `UDP` (pour l'envoie des messages) et `HTTP` (pour l'export de statistiques) ainsi que les noms de nos `queues` et les identifiants des `processor` associés. Nous reviendrons sur les processors un peu plus tard.
+
+Ce qui est intéressant est que nous allons également pouvoir contrôler ici est, pour chaque `queue`, le nombre de workers que nous souhaitons voir disponible.
+
+Côté Go, nous avons commencés par importer les librairies nécessaires, ici, que des packages natifs, ce qui m'a permis d'expliquer rapidement le concept d'import de librairies internes mais aussi externes ainsi que que la logique de namespaces en répertoires dans un projet :
```
Comment (ch3ric) [inline _posts/2017-08-23-retour-sur-un-live-coding-de-decouverte-du-langage-go.md]: commencé

File: _posts/2017-08-23-retour-sur-un-live-coding-de-decouverte-du-langage-go.md [hunk 9]
```diff
@@ -0,0 +1,366 @@
...
+	return &Workmq{
+		Config:     config,
+		Queues:     queues,
+		Processors: processors,
+		Counters:   counters,
+	}
+}
+```
+
+Dans ce bout de code qui initialise un pointeur de structure `Workmq`, la plupart des sujets que je souhaitais traiter son à expliquer :
```
Comment (ch3ric) [inline _posts/2017-08-23-retour-sur-un-live-coding-de-decouverte-du-langage-go.md]: sont

File: _posts/2017-08-23-retour-sur-un-live-coding-de-decouverte-du-langage-go.md [hunk 10]
```diff
@@ -0,0 +1,366 @@
...
+
+	err := http.ListenAndServe(w.Config.Ports.HTTP, nil)
+
+	if err != nil {
+		log.Fatal("ListenAndServe error: ", err)
+	}
+}
+```
+
+Nous avons ici bouclés sur toutes les queues et affichons les données des compteurs en sortie. Rien de vraiment sorcier par rapport à ce que nous avons déjà appris précédemment.
```
Comment (ch3ric) [inline _posts/2017-08-23-retour-sur-un-live-coding-de-decouverte-du-langage-go.md]: bouclé

**Comment by eko** [comment]
Thanks, the spelling mistakes are fixed :)

**Pull request merged**